diff -u -r ata-orig/ata-lowlevel.c ata/ata-lowlevel.c
--- ata-orig/ata-lowlevel.c	2006-06-16 14:32:11.000000000 -0700
+++ ata/ata-lowlevel.c	2006-06-16 16:10:51.000000000 -0700
@@ -33,6 +33,7 @@
 #include <sys/param.h>
 #include <sys/systm.h>
 #include <sys/ata.h>
+#include <sys/bio.h>
 #include <sys/kernel.h>
 #include <sys/conf.h>
 #include <sys/bus.h>
@@ -141,6 +142,15 @@
 	}
 
 	/* issue command */
+	if (strcmp(request->device->name, "ad1") &&
+	    strcmp(request->device->name, "ad4")) {
+		printf("write: lba:%d, dev:'%s', cnt:%d, bs:%d,%d\n",
+			(int)request->u.ata.lba,
+			(char*)request->device->name,
+			(int)request->u.ata.count,
+			(int)request->bytecount,
+			(int)request->transfersize);
+	}
 	if (ch->hw.command(request->device, request->u.ata.command,
 			   request->u.ata.lba, request->u.ata.count,
 			   request->u.ata.feature)) {
