diff -u linux-2.6.12/drivers/Kconfig linux-2.6.12/drivers/Kconfig
--- linux-2.6.12/drivers/Kconfig	2005-06-17 12:48:29.000000000 -0700
+++ linux-2.6.12/drivers/Kconfig	2006-04-03 16:06:56.000000000 -0700
@@ -4,6 +4,8 @@
 
 source "drivers/base/Kconfig"
 
+source "drivers/fstitch/Kconfig"
+
 source "drivers/mtd/Kconfig"
 
 source "drivers/parport/Kconfig"
diff -u linux-2.6.12/drivers/Makefile linux-2.6.12/drivers/Makefile
--- linux-2.6.12/drivers/Makefile	2005-06-17 12:48:29.000000000 -0700
+++ linux-2.6.12/drivers/Makefile	2006-04-03 16:09:21.000000000 -0700
@@ -17,6 +17,8 @@
 # default.
 obj-y				+= char/
 
+obj-$(CONFIG_FSTITCH_PROC)	+= fstitch/
+
 # i810fb and intelfb depend on char/agp/
 obj-$(CONFIG_FB_I810)           += video/i810/
 obj-$(CONFIG_FB_INTEL)          += video/intelfb/
diff -u linux-2.6.12/drivers/fstitch/Kconfig linux-2.6.12/drivers/fstitch/Kconfig
--- linux-2.6.12/drivers/fstitch/Kconfig	1969-12-31 16:00:00.000000000 -0800
+++ linux-2.6.12/drivers/fstitch/Kconfig	2006-04-03 16:01:11.000000000 -0700
@@ -0,0 +1,5 @@
+config FSTITCH_PROC
+	boolean "Provide process event hooks for Featherstitch"
+	---help---
+	  This option provides hooks for the Featherstitch kernel module to find out
+          about process events, which it needs to support patchgroups.
diff -u linux-2.6.12/drivers/fstitch/Makefile linux-2.6.12/drivers/fstitch/Makefile
--- linux-2.6.12/drivers/fstitch/Makefile	1969-12-31 16:00:00.000000000 -0800
+++ linux-2.6.12/drivers/fstitch/Makefile	2006-04-03 16:12:41.000000000 -0700
@@ -0,0 +1 @@
+obj-$(CONFIG_FSTITCH_PROC)	+= fstitch_proc.o
diff -u linux-2.6.12/drivers/fstitch/fstitch_proc.c linux-2.6.12/drivers/fstitch/fstitch_proc.c
--- linux-2.6.12/drivers/fstitch/fstitch_proc.c	1969-12-31 16:00:00.000000000 -0800
+++ linux-2.6.12/drivers/fstitch/fstitch_proc.c	2006-04-04 01:43:28.000000000 -0700
@@ -0,0 +1,102 @@
+/*
+ * fstitch_proc.c - provide process event hooks for Featherstitch
+ *
+ * Copyright (C) 2006 Regents of the University of California
+ * Based loosely on cn_proc.c by Matt Helsley
+ * 
+ * Copyright (C) 2005 Matt Helsley, IBM Corp.
+ * Based on cn_fork.c by Guillaume Thouvenin <guillaume.thouvenin@bull.net>
+ * Original copyright notice follows:
+ * Copyright (C) 2005 BULL SA.
+ *
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ */
+
+#include <linux/module.h>
+#include <linux/kernel.h>
+#include <linux/init.h>
+#include <asm/atomic.h>
+
+#include <linux/fstitch_proc.h>
+
+static atomic_t fstitch_module_loaded = ATOMIC_INIT(0);
+static spinlock_t fstitch_spinlock = SPIN_LOCK_UNLOCKED;
+static struct fstitch_proc_ops * fstitch_module = NULL;
+
+void fstitch_fork_hook(struct task_struct * task)
+{
+	if(atomic_read(&fstitch_module_loaded) < 1)
+		return;
+	
+	fstitch_module->fork(task);
+}
+
+void fstitch_exec_hook(struct task_struct * task)
+{
+	if(atomic_read(&fstitch_module_loaded) < 1)
+		return;
+	
+	fstitch_module->exec(task);
+}
+
+void fstitch_exit_hook(struct task_struct * task)
+{
+	if(atomic_read(&fstitch_module_loaded) < 1)
+		return;
+	
+	fstitch_module->exit(task);
+}
+
+int fstitch_register_module(struct fstitch_proc_ops * ops)
+{
+	spin_lock(&fstitch_spinlock);
+	
+	if(atomic_read(&fstitch_module_loaded) > 0)
+	{
+		spin_unlock(&fstitch_spinlock);
+		return -EBUSY;
+	}
+	
+	fstitch_module = ops;
+	
+	atomic_inc(&fstitch_module_loaded);
+	
+	spin_unlock(&fstitch_spinlock);
+	
+	return 0;
+}
+
+int fstitch_unregister_module(struct fstitch_proc_ops * ops)
+{
+	spin_lock(&fstitch_spinlock);
+	
+	if(atomic_read(&fstitch_module_loaded) < 1 || fstitch_module != ops)
+	{
+		spin_unlock(&fstitch_spinlock);
+		return -EINVAL;
+	}
+	
+	atomic_dec(&fstitch_module_loaded);
+	
+	fstitch_module = NULL;
+	
+	spin_unlock(&fstitch_spinlock);
+	
+	return 0;
+}
+
+EXPORT_SYMBOL(fstitch_register_module);
+EXPORT_SYMBOL(fstitch_unregister_module);
diff -u linux-2.6.12/fs/exec.c linux-2.6.12/fs/exec.c
--- linux-2.6.12/fs/exec.c	2006-03-11 08:12:04.000000000 -0800
+++ linux-2.6.12/fs/exec.c	2006-04-04 00:56:06.000000000 -0700
@@ -48,6 +48,7 @@
 #include <linux/syscalls.h>
 #include <linux/rmap.h>
 #include <linux/acct.h>
+#include <linux/fstitch_proc.h>
 
 #include <asm/uaccess.h>
 #include <asm/mmu_context.h>
@@ -1091,6 +1092,7 @@
 					fput(bprm->file);
 				bprm->file = NULL;
 				current->did_exec = 1;
+				fstitch_exec_hook(current);
 				return retval;
 			}
 			read_lock(&binfmt_lock);
diff -u linux-2.6.12/include/linux/fstitch_proc.h linux-2.6.12/include/linux/fstitch_proc.h
--- linux-2.6.12/include/linux/fstitch_proc.h	1969-12-31 16:00:00.000000000 -0800
+++ linux-2.6.12/include/linux/fstitch_proc.h	2006-04-04 01:43:34.000000000 -0700
@@ -0,0 +1,61 @@
+/*
+ * fstitch_proc.h - provide process event hooks for Featherstitch
+ *
+ * Copyright (C) 2006 Regents of the University of California
+ * Based loosely on cn_proc.h by Matt Helsley
+ * 
+ * Copyright (C) Matt Helsley, IBM Corp. 2005
+ * Based on cn_fork.h by Nguyen Anh Quynh and Guillaume Thouvenin
+ * Original copyright notice follows:
+ * Copyright (C) 2005 Nguyen Anh Quynh <aquynh@gmail.com>
+ * Copyright (C) 2005 Guillaume Thouvenin <guillaume.thouvenin@bull.net>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ */
+
+#ifndef FSTITCH_PROC_H
+#define FSTITCH_PROC_H
+
+#include <linux/config.h>
+
+#ifdef __KERNEL__
+
+#ifdef CONFIG_FSTITCH_PROC
+
+struct task_struct;
+
+struct fstitch_proc_ops {
+	void (*fork)(struct task_struct * child);
+	void (*exec)(struct task_struct * process);
+	void (*exit)(struct task_struct * process);
+};
+
+void fstitch_fork_hook(struct task_struct * task);
+void fstitch_exec_hook(struct task_struct * task);
+void fstitch_exit_hook(struct task_struct * task);
+
+int fstitch_register_module(struct fstitch_proc_ops * ops);
+int fstitch_unregister_module(struct fstitch_proc_ops * ops);
+
+#else
+
+static inline void fstitch_fork_hook(struct task_struct * task) {}
+static inline void fstitch_exec_hook(struct task_struct * task) {}
+static inline void fstitch_exit_hook(struct task_struct * task) {}
+
+#endif	/* CONFIG_FSTITCH_PROC */
+
+#endif	/* __KERNEL__ */
+#endif	/* FSTITCH_PROC_H */
diff -u linux-2.6.12/kernel/exit.c linux-2.6.12/kernel/exit.c
--- linux-2.6.12/kernel/exit.c	2006-03-11 08:12:12.000000000 -0800
+++ linux-2.6.12/kernel/exit.c	2006-04-04 00:57:59.000000000 -0700
@@ -28,6 +28,7 @@
 #include <linux/cpuset.h>
 #include <linux/syscalls.h>
 #include <linux/signal.h>
+#include <linux/fstitch_proc.h>
 
 #include <asm/uaccess.h>
 #include <asm/unistd.h>
@@ -832,6 +833,7 @@
 		module_put(tsk->binfmt->module);
 
 	tsk->exit_code = code;
+	fstitch_exit_hook(tsk);
 	exit_notify(tsk);
 #ifdef CONFIG_NUMA
 	mpol_free(tsk->mempolicy);
diff -u linux-2.6.12/kernel/fork.c linux-2.6.12/kernel/fork.c
--- linux-2.6.12/kernel/fork.c	2005-06-17 12:48:29.000000000 -0700
+++ linux-2.6.12/kernel/fork.c	2006-04-04 00:53:35.000000000 -0700
@@ -41,6 +41,7 @@
 #include <linux/profile.h>
 #include <linux/rmap.h>
 #include <linux/acct.h>
+#include <linux/fstitch_proc.h>
 
 #include <asm/pgtable.h>
 #include <asm/pgalloc.h>
@@ -1102,6 +1103,7 @@
 	nr_threads++;
 	total_forks++;
 	write_unlock_irq(&tasklist_lock);
+	fstitch_fork_hook(p);
 	retval = 0;
 
 fork_out:
