--- kernel/softlockup.c	2007-02-19 22:34:32.000000000 -0800
+++ kernel/softlockup.c	2007-03-28 16:03:54.000000000 -0700
@@ -13,6 +13,7 @@
 #include <linux/kthread.h>
 #include <linux/notifier.h>
 #include <linux/module.h>
+#include <linux/reboot.h>
 
 static DEFINE_SPINLOCK(print_lock);
 
@@ -67,12 +68,22 @@
 
 	/* Warn about unreasonable 10+ seconds delays: */
 	if (time_after(jiffies, touch_timestamp + 10*HZ)) {
+		static int tries = 0;
+		static unsigned long expire = 0;
 		per_cpu(print_timestamp, this_cpu) = touch_timestamp;
 
 		spin_lock(&print_lock);
 		printk(KERN_ERR "BUG: soft lockup detected on CPU#%d!\n",
 			this_cpu);
 		dump_stack();
+		if (jiffies - expire > 60 * HZ) {
+			expire = jiffies;
+			tries = 0;
+		}
+		if (++tries >= 6) {
+			printk(KERN_EMERG "Soft lockup is too long, rebooting!\n");
+			emergency_restart();
+		}
 		spin_unlock(&print_lock);
 	}
 }
